### PR TITLE
Add Busola migrator patch job

### PR DESCRIPTION
**Pull request by dbadura**
Rewrite the virtual service migrator script for busola to `Go`.

Issue: https://github.com/kyma-project/kyma/issues/11795

### CODE REVIEW

File: pkg/reconciler/instances/busola/test_files/virtual-service.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    "labels": {
+      "app": "dex",
+      "app.kubernetes.io/managed-by": "Helm",
+      "chart": "dex-1.0.0"
+    },
+    "managedFields": [
+      {
+        "apiVersion": "networking.istio.io/v1alpha3",
+        "fieldsType": "FieldsV1",
+        "fieldsV1": {
```
Comment (dbadura) [inline pkg/reconciler/instances/busola/test_files/virtual-service.yaml:20]: I think that I should remove all those empty objects

File: pkg/reconciler/kubernetes/adapter/adapter.go [hunk 2]
```diff
@@ -151,6 +151,7 @@ func (g *kubeClientAdapter) deployManifest(ctx context.Context, manifest, namesp
 			resource, err := g.kubeClient.ApplyWithNamespaceOverride(&unstruct, namespace)
 			if err != nil {
 				g.logger.Errorf("Failed to apply Kubernetes unstructured entity: %s", err)
+				g.logger.Debugf("Used JSON data: %+v", unstruct)
```
Comment (dbadura) [inline pkg/reconciler/kubernetes/adapter/adapter.go:154]: To remove

File: pkg/reconciler/instances/busola/action.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	clientSet, err := helper.KubeClient.Clientset()
+	if err != nil {
+		return errors.Wrapf(err, "while getting client set from kubeclient")
+	}
+	restClient := clientSet.Discovery().RESTClient()
+
+	logger.Infof("Launching pre install busola migrator job, version: %s ", version)
+	for _, virtSvcToPatch := range p.virtSvcsToPatch {
+		logger.Infof("Patching virtual service: %s in namespace: %s", virtSvcToPatch.Name, virtSvcToPatch.Namespace)
+		if err := p.patchVirtSvc(ctx, restClient, virtSvcToPatch.Name, virtSvcToPatch.Namespace, logger); err != nil {
```
Comment (tgorgol) [inline pkg/reconciler/instances/busola/action.go:54]: What will happen if the virtual service doesn't exist?

File: pkg/reconciler/instances/busola/busola.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+	//configure reconciler
+	reconciler.
+		//list dependencies (these components have to be available before this component reconciler is able to run)
+		WithDependencies("istio", "dex", "console").
+		//register reconciler pre-action (executed BEFORE reconciliation happens)
+		WithPreReconcileAction(&VirtualServicePreInstallPatch{
+			name:            "pre-action",
+			virtSvcsToPatch: virtSvcs,
+			suffix:          "-old",
```
Comment (tgorgol) [inline pkg/reconciler/instances/busola/busola.go:43]: These values used to be configured via helm values, do we want to hardcode them?

**Review by tgorgol — COMMENTED** [review]
Few comments from me

File: pkg/reconciler/instances/loader.go [hunk 5]
```diff
@@ -5,6 +5,8 @@ package instances
 import (
 	//import required to register component reconciler 'base' in reconciler registry
 	_ "github.com/kyma-incubator/reconciler/pkg/reconciler/instances/base"
+	//import required to register component reconciler 'busola' in reconciler registry
+	_ "github.com/kyma-incubator/reconciler/pkg/reconciler/instances/busola"
```
[the remark this answers is not in the document]
Comment (tobiscr) [reply pkg/reconciler/instances/loader.go:9]: The package name is not restricted or has to follow a naming convention. You are free to choose any name you like.

File: pkg/reconciler/instances/busola/vs_client.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+func NewVirtSvcClient() VirtSvcClient {
+	return &client{}
+}
+
+const (
+	istioCRPath  = "/apis/networking.istio.io/v1alpha3"
+	resourceName = "virtualservices"
+)
+
+type virtSvc struct {
```
[the remark this answers is not in the document]
Comment (tobiscr) [reply pkg/reconciler/instances/busola/vs_client.go:30]: Please have in mind, that adding further APIs increases also the surface for potential security issues. You are of course not limited to use only APIs we have in place and it's ok to add furhter APIs. But please have in the amount of additional dependencies in mind.. the ISTIO client has a lot of them: https://github.com/istio/client-go/blob/master/go.sum

File: pkg/reconciler/instances/loader.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbadura) [reply pkg/reconciler/instances/loader.go:9]: I renamed the pacakge to `busolamigrator`

File: pkg/reconciler/instances/busola/busola.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+		{
+			Name:      "dex-virtualservice",
+			Namespace: "kyma-system",
+		},
+	}
+
+	//configure reconciler
+	reconciler.
+		//list dependencies (these components have to be available before this component reconciler is able to run)
+		WithDependencies("istio", "dex", "console").
```
[the remark this answers is not in the document]
Comment (dbadura) [reply pkg/reconciler/instances/busola/busola.go:38]: As `Virtual Service` is need, we only need to check for `istio` dependency. 
I removed the old ones.

File: pkg/reconciler/instances/busola/action.go (hunk 3 above)
Comment (dbadura) [reply pkg/reconciler/instances/busola/action.go:54]: There will be an error and the step will fail.
I added test case for this.

File: pkg/reconciler/instances/loader.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tobiscr) [reply pkg/reconciler/instances/loader.go:9]: 👍

**Review by tobiscr — DISMISSED** [review]

File: pkg/reconciler/instances/busola/action.go (hunk 3 above)
Comment (tgorgol) [reply pkg/reconciler/instances/busola/action.go:54]: Then it behaves differently than the current helm hook implementation and having it work without dex/console components installed was one of the requirements if I remember correctly.

Comment (dbadura) [reply pkg/reconciler/instances/busola/action.go:54]: You're right. I am going to fix it

File: pkg/reconciler/instances/busola/busola.go (hunk 4 above)
Comment (dbadura) [reply pkg/reconciler/instances/busola/busola.go]: Are we going to persinalize it?
I think no.

File: pkg/reconciler/instances/busola/vs_client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbadura) [reply pkg/reconciler/instances/busola/vs_client.go]: Busola migrator is simple job, so we can go with rest client.

[the remark this answers is not in the document]
Comment (tobiscr) [reply pkg/reconciler/instances/busola/vs_client.go]: 👍

**Review by tgorgol — APPROVED** [review]
LGTM

**Review by tobiscr — APPROVED** [review]

**Pull request merged**